File: examples/ch15/ts/conceptVisitor.ts
// Visitor — canonical conceptual example. Run: npx tsx conceptVisitor.ts
//
// Double dispatch: element.accept(visitor) calls back visitor.visitX(this). The
// modern replacement is a discriminated union + exhaustive switch.

export interface Visitor<T> {
  visitA(element: ElementA): T;
  visitB(element: ElementB): T;
}

export interface Element {
  accept<T>(visitor: Visitor<T>): T;
}

export class ElementA implements Element {
  accept<T>(visitor: Visitor<T>): T {
    return visitor.visitA(this);
  }
}

export class ElementB implements Element {
  accept<T>(visitor: Visitor<T>): T {
    return visitor.visitB(this);
  }
}

export class NameVisitor implements Visitor<string> {
  visitA(_element: ElementA): string {
    return "A";
  }
  visitB(_element: ElementB): string {
    return "B";
  }
}

// Demo
const elements: Element[] = [new ElementA(), new ElementB()];
console.log(elements.map((e) => e.accept(new NameVisitor()))); // ['A', 'B']

File: examples/ch15/py/concept_visitor.py
"""Visitor — canonical conceptual example (runnable).

Double dispatch: element.accept(visitor) calls back visitor.visit_x(self), so
the right operation runs for the right type. The modern replacement is a single
function using structural match (see total_via_match in visitor.py).
"""

from __future__ import annotations

from abc import ABC, abstractmethod


class Visitor(ABC):
    @abstractmethod
    def visit_a(self, element: ElementA) -> str: ...

    @abstractmethod
    def visit_b(self, element: ElementB) -> str: ...


class Element(ABC):
    @abstractmethod
    def accept(self, visitor: Visitor) -> str: ...


class ElementA(Element):
    def accept(self, visitor: Visitor) -> str:
        return visitor.visit_a(self)


class ElementB(Element):
    def accept(self, visitor: Visitor) -> str:
        return visitor.visit_b(self)


class NameVisitor(Visitor):
    def visit_a(self, element: ElementA) -> str:
        return "A"

    def visit_b(self, element: ElementB) -> str:
        return "B"


if __name__ == "__main__":
    elements: list[Element] = [ElementA(), ElementB()]
    print([e.accept(NameVisitor()) for e in elements])  # ['A', 'B']
